Userland/Libraries/LibGUI/Label.cpp

```diff
@@ -161,28 +181,83 @@ void Label::wrap_text()
         rect.set_width(rect.width() - font().glyph_width('x'));
 
     Vector<String> lines;
+    String line;
     StringBuilder builder;
-    int line_width = 0;
     for (auto& word : words) {
-        int word_width = font().width(word);
-        if (line_width > 0)
-            word_width += font().glyph_width('x');
-        if (line_width + word_width > rect.width()) {
-            lines.append(builder.to_string());
-            builder.clear();
-            line_width = 0;
+        switch (m_word_wrap_mode) {
+        case GUI::WordWrapMode::Word:
+            line = builder.to_string();
+            builder.append(word);
+            if (font().width(builder.to_string()) > rect.width() && !line.is_empty()) {
+                lines.append(line);
+                line = String::empty();
+                builder.clear();
+                builder.append(word);
+            }
+            if (word == "\n") {
+                if (!line.is_empty())
+                    lines.append(line);
+                builder.clear();
+                lines.append(word);
+            }
+            continue;
+        case GUI::WordWrapMode::Anywhere:
+            size_t current_position = 0;
+            while (current_position != word.length()) {
+                size_t size = next_substring_size(word, current_position);
+                lines.append(word.substring(current_position, size));
+                current_position += size;
+            }
+            continue;
         }
-        if (line_width > 0)
-            builder.append(' ');
-        builder.append(word);
-        line_width += word_width;
     }
 
-    auto last_line = builder.to_string();
-    if (!last_line.is_empty())
-        lines.append(last_line);
+    if (!builder.is_empty())
+        lines.append(builder.to_string());
 
     m_lines = lines;
 }
 
+size_t Label::next_substring_size(String word, size_t start)
+{
+    String sub_word = word.substring(start);
+    size_t lower_idx = 0, upper_idx = sub_word.length();
+    size_t idx = (upper_idx - lower_idx) / 2;
```

Comment:
Please avoid the [`idx` shortening](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#names), we're not on a 80-character limit :^)